doc/dev/transforms.rst

```diff
@@ -100,6 +100,117 @@ classical instructions as well as side-effecting operations (e.g., print) can al
+
+The first thing to do is to create the pass object in the tablegen ``mlir/include/Catalyst/Transforms/Passes.td``:
+
+.. code-block::
+
+    def MyHelloWorldPass : Pass<"my-hello-world"> {
+        let summary = "An empty pass boilerplate that prints out hello world.";
+
+        let constructor = "catalyst::createMyHelloWorldPass()";
+    }
+
+When the dialect is built, this tablegen def will be built to a C++ file ``Catalyst/Transforms/Passes.h.inc``, containing the newly defined object called ``MyHelloWorldPass``, alongside the various necessary boilerplate methods in the MLIR infrastructure. Tablegen is designed such that we don't have to write all that boilerplate ourselves. 
+
+Now we write the pass itself. Create a new file ``mlir/lib/Catalyst/Transforms/MyHelloWorldPass.cpp`` with the following content:
+
+.. code-block:: cpp
+
+    #define DEBUG_TYPE "myhelloworld"
+
+    #include "Catalyst/IR/CatalystDialect.h"
+    #include "mlir/Pass/Pass.h"
+    #include "llvm/Support/Debug.h"
+
+    using namespace llvm;
+    using namespace mlir;
+    using namespace catalyst;
+
+    namespace catalyst {
+    #define GEN_PASS_DEF_MYHELLOWORLDPASS
+    #include "Catalyst/Transforms/Passes.h.inc"
+
+    struct MyHelloWorldPass : public impl::MyHelloWorldPassBase<MyHelloWorldPass> {
+        using impl::MyHelloWorldPassBase<MyHelloWorldPass>::MyHelloWorldPassBase;
+
+        void runOnOperation() override { llvm::errs() << "Hello world!\n"; }
+    };
+
+    std::unique_ptr<Pass> createMyHelloWorldPass() { return std::make_unique<MyHelloWorldPass>(); }
+
+    } // namespace catalyst
+
+The function that determines what your pass actually does is the ``void runOnOperation()``. Here all the pass does is print out ``"Hello world!\n"``. 
+
+This new C++ file needs to be added to the ``mlir/lib/Catalyst/Transforms/CMakeLists.txt`` file (or the CMakeLists.txt of whichever directory that has your new pass file): 
+
+.. code-block::
+
+    file(GLOB SRC
+        ...
+        MyHelloWorldPass.cpp
+    )
+
+After writing the pass, we need to register it in a few places. In ``mlir/include/Catalyst/Transforms/Passes.h``, add the method 
+
+.. code-block:: cpp
+
+    namespace catalyst {
+        ...
+        std::unique_ptr<mlir::Pass> createMyHelloWorldPass();
+        ...
+    }
+
+And in ``mlir/lib/Catalyst/Transforms/RegisterAllPasses.cpp``, register the pass via 
+
+.. code-block:: cpp
+
+    void catalyst::registerAllCatalystPasses()
+    {
+        ...
+        mlir::registerPass(catalyst::createMyHelloWorldPass);
+        ...
+    }
+
+Now that we have written our shiny new pass, we can build it by going back to the top-level ``catalyst`` directory and 
+
+.. code-block::
+
+    make dialects
+
+The tool to run passes is built as ``mlir/build/bin/quantum-opt``.
+
+We can inspect by all the available passes by running ``quantum-opt --help``:
```

Comment:
I need to either navigate to `mlir/build/bin/` and then run `./quantum-opt --help` or directly run `./mlir/build/bin/quantum-opt --help`, otherwise it doesnt work. Is that macOS specific and clear for everyone else? Not sure I would have gotten that if @dime10 didnt show me how to use the tool somewhere else

Reply:
This is how to run executables for any linux-style terminals, including the one on mac. It's the same issue where `a.out` will not run a compiled C program but `./a.out` will. I will add this in the docs as well